dialects/linux/dproc.c

```diff
@@ -1518,12 +1574,91 @@ process_proc_map(p, s, ss)
 		efs = sv = 1;
 	    else
 		efs = 0;
-	    if (!efs) {
-		if (HasNFS)
-		    sv = statsafely(fp[6], &sb);
-		else
-		    sv = stat(fp[6], &sb);
+
```

Comment:
I don't understand why you added your code that introduces extra overhead here.

Isn't it enough to run your code only when `stat` is failed?

Reply:
I understand the situation where this is not so simple.

I think the following algorithm is better.

1. get a line from /proc/$pid/maps. Let's call it $line.
2. get a file name from the $line. Let's call it $fname.
3. Call stat($fname). Let's call the result of stat buf $stat. If the syscall returns an error, let's use your new code.
4. get a (major,minor,inode) tuple from $line. Let's call the tuple $INODE.
5. compare $INODE with $stat. If they don't match, let's use your new code.
6. we confirmed $stat includes enough useful information. We don't have to run your new code.

Using fork & pipe takes costs. If just using stat is enough, we don't have to take the costs.

Reply:
This algorithm has less intrusion to current code, but I think logically, it seems not better.
According to this steps, we should always call stat(2) first of all, when we get failed, turn to the new code to try again. But **we can know it will surely fail** by namespace comparison in advance, so we don't need to call stat(2) firstly, and thus, cost of this syscall is saved.
About fork and pipe costs, this patch mainly obtain correct stat result at the expense of these costs for the processes in different mnt namespace indeed. For others, the stat() is called directly, thus the costs are not taken.

Reply:
> About fork and pipe costs, this patch mainly obtain correct stat result at the expense of these costs for the processes in different mnt namespace indeed. For others, the stat() is called directly, thus the costs are not taken.

(with your pull request + close(pipefs[0]) patch):
```
$ sudo time ./lsof > /dev/null
...(stderr)...
7.83user 318.46system 5:32.45elapsed 98%CPU (0avgtext+0avgdata 433404maxresident)k
8inputs+0outputs (0major+1565450minor)pagefaults 0swaps
$ git diff                    
diff --git a/dialects/linux/dproc.c b/dialects/linux/dproc.c
index bf6b216..8fb96e8 100644
--- a/dialects/linux/dproc.c
+++ b/dialects/linux/dproc.c
@@ -1658,6 +1658,7 @@ process_proc_map(p, s, ss)
                        goto stat_directly;
                    }
 
+                   close(pipefd[0]);
                    /* Wait the child process exit */
                    waitpid(pid, &status, 0);
```

Without your pull request:
```
$ git checkout master                                                     
Switched to branch 'master'                                                                                                                                                                   Your branch is up to date with 'origin/master'.                      
$ make -j 10
...
$ sudo time ./lsof > /dev/null
...(stderr)...
6.18user 7.35system 0:17.08elapsed 79%CPU (0avgtext+0avgdata 440348maxresident)k
0inputs+0outputs (0major+109423minor)pagefaults 0swaps
```

Reply:
That's a large performance regression. Can we compare the ns in a faster way? Or we will have to use the fallback approach.

Reply:
Can we move one process to another mount namespace and then move it back?

Reply:
```
$ uname -a
Linux living.shokupan 5.18.11-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Tue Jul 12 22:52:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ sudo lsns -t mnt                                   
        NS TYPE NPROCS    PID USER            COMMAND
4026531841 mnt     596      1 root            /usr/lib/systemd/systemd --switched-root --system --deserialize 35
4026532188 mnt       1    754 root            ├─/usr/lib/systemd/systemd-udevd
4026532462 mnt       2    980 root            ├─/sbin/auditd
4026532464 mnt       1    978 systemd-oom     ├─/usr/lib/systemd/systemd-oomd
4026532465 mnt       1    985 systemd-resolve ├─/usr/lib/systemd/systemd-resolved
4026532475 mnt       2   1010 dbus            ├─/usr/bin/dbus-broker-launch --scope system --audit
4026532476 mnt       1   1018 root            ├─/usr/libexec/bluetooth/bluetoothd
4026532477 mnt       1   1061 chrony          ├─/usr/sbin/chronyd -F 2
4026532545 mnt       1   1053 root            ├─gpm -m /dev/input/mice -t exps2
4026532613 mnt       1   1026 root            ├─/usr/libexec/low-memory-monitor
4026532614 mnt       1   1036 root            ├─/usr/libexec/power-profiles-daemon
4026532682 mnt       1   1039 rtkit           ├─/usr/libexec/rtkit-daemon
4026532683 mnt       1   1040 root            ├─/usr/libexec/switcheroo-control
4026532686 mnt       1   1042 root            ├─/usr/lib/systemd/systemd-logind
4026532757 mnt       1   1106 root            ├─/usr/sbin/ModemManager
4026532758 mnt       1   1115 root            ├─/usr/sbin/NetworkManager --no-daemon
4026532759 mnt       6   1137 root            ├─php-fpm: master process (/etc/php-fpm.conf)
4026532784 mnt       5   1282 root            ├─/usr/sbin/httpd -DFOREGROUND
4026532786 mnt       1   2920 root            ├─/usr/libexec/upowerd
4026532790 mnt       1   4569 colord          └─/usr/libexec/colord
4026531862 mnt       1     74 root            kdevtmpfs
4026534282 mnt       1 208843 jet             catatonit -P
4026535188 mnt       1 209112 root            /bin/bash
```

Reply:
It seems that fork is much more expensive than stat.

Reply:
With this pull request:
```
$ sudo strace -f -c -e clone ./lsof > /dev/null
...
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
100.00  123.704179        5040     24540           clone
------ ----------- ----------- --------- --------- ----------------
100.00  123.704179        5040     24540           total
```
Improving the performance of clone() is good for the human being :-P
However, I guess there is no space for improvement.

You created a process for calling stat in a namespace repeatedly.
For a namespace, we need only one process running background.
For N namespace, we need only N processes. Running the process as a server.